google-ads-examples/src/main/java/com/google/ads/googleads/examples/planning/GenerateKeywordIdeas.java

```diff
@@ -0,0 +1,202 @@
+              params.languageId,
+              params.locationIds,
+              params.keywords,
+              params.pageUrl);
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param languageId the language ID.
+   * @param locationIds the location IDs.
+   * @param keywords the list of keywords to use as a seed for ideas.
+   * @param pageUrl optional URL related to your business to use as a seed for ideas.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   * @throws IllegalArgumentException if {@code keywords} is empty and {@code pageUrl} is null.
+   * @throws Exception if the example failed due to other errors.
+   */
+  private void runExample(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      long languageId,
+      List<Long> locationIds,
+      List<String> keywords,
+      @Nullable String pageUrl) {
+    try (KeywordPlanIdeaServiceClient keywordPlanServiceClient =
+        googleAdsClient.getKeywordPlanIdeaServiceClient()) {
+      GenerateKeywordIdeasRequest.Builder requestBuilder =
+          GenerateKeywordIdeasRequest.newBuilder()
+              .setCustomerId(Long.toString(customerId))
+              // Set the language resource using the provided language ID.
+              .setLanguage(StringValue.of(ResourceNames.languageConstant(languageId)))
+              // Set the network. To restrict to only Google Search, change the parameter below to
+              // KeywordPlanNetwork.GOOGLE_SEARCH.
+              .setKeywordPlanNetwork(KeywordPlanNetwork.GOOGLE_SEARCH_AND_PARTNERS);
+
+      // Add the resource name of each location ID to the request.
+      for (Long locationId : locationIds) {
+        requestBuilder.addGeoTargetConstants(
+            StringValue.of(ResourceNames.geoTargetConstant(locationId)));
+      }
+
+      // Make sure that keywords and/or page URL were specified. Otherwise, there are no seed values
+      // for the request.
+      if (keywords.isEmpty() && pageUrl == null) {
+        throw new IllegalArgumentException(
+            "At least one of keywords or page URL is required, but neither was specified.");
+      }
+
+      if (keywords.isEmpty()) {
```

Comment:
You could also add the above error branch to bring the entire oneof handling into a single block.

Reply:
Reworked to reduce nesting. I did not include the error case here because if that's true, the example will throw an exception. I personally prefer separating out the error case, but LMK if you feel strongly about combining them.

Reply:
Yeah that's a good point, thanks.